homeassistant/components/xiaomi_miio/fan.py

```diff
@@ -1087,17 +1099,21 @@ async def async_set_percentage(self, percentage: int) -> None:
             await self.async_turn_off()
             return
 
+        speed = math.ceil(
+            percentage_to_ranged_value((1, self._speed_count), percentage)
+        )
+
+        # if the fan is not on, we have to turn it on first
+        if not self.is_on:
+            await self.async_turn_on()
+
         await self._try_command(
             "Setting fan speed percentage of the miio device failed.",
             self._device.set_speed,
-            percentage,
+            speed,
         )
         self._percentage = percentage
-
-        if not self.is_on:
-            await self.async_turn_on()
-        else:
-            self.async_write_ha_state()
+        self.async_write_ha_state()
```

Comment:
Maybe only update state if the api call was successful and returned True, like is done in `async_turn_on`?

Reply:
I changed it accordingly